domain/removal/state/unit.go

```diff
@@ -65,62 +68,101 @@ AND    life_id = 0`, unitUUID)
+), machines AS (
+    SELECT    m.uuid AS machine_uuid,
+              m.net_node_uuid,
+              COUNT(ua.uuid) AS unit_alive_count,
+              COUNT(una.uuid) AS unit_not_alive_count
+    FROM      machine AS m
+    JOIN      net_node AS nn ON nn.uuid = m.net_node_uuid
+    LEFT JOIN units_alive AS ua ON ua.net_node_uuid = nn.uuid
+    LEFT JOIN units_not_alive AS una ON una.net_node_uuid = nn.uuid
+    GROUP BY  m.uuid
 )
-SELECT unit_count AS &entityAssociationCount.count,
-	   machine_uuid AS &entityAssociationCount.uuid
+SELECT unit_alive_count AS &entityAssociationAliveCount.alive_count,
+       unit_not_alive_count AS &entityAssociationAliveCount.not_alive_count,
+       machine_uuid AS &entityAssociationAliveCount.uuid
 FROM   machines
-WHERE  unit_uuid = $entityUUID.uuid;
-	`, unitUUID, entityAssociationCount{})
+LEFT JOIN unit AS u ON u.net_node_uuid = machines.net_node_uuid
+WHERE  u.uuid = $entityUUID.uuid;
+    `, unitUUID, entityAssociationAliveCount{})
 	if err != nil {
 		return "", errors.Errorf("preparing unit count query: %w", err)
 	}
 
 	updateMachineStmt, err := st.Prepare(`
 UPDATE machine
 SET    life_id = 1
-WHERE  uuid = $entityAssociationCount.uuid
-AND    life_id = 0`, entityAssociationCount{})
+WHERE  uuid = $entityUUID.uuid
+AND    life_id = 0`, entityUUID{})
 	if err != nil {
 		return "", errors.Errorf("preparing machine life update: %w", err)
 	}
 
-	var mUUID string
-	if err := errors.Capture(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		if err := tx.Query(ctx, updateUnitStmt, unitUUID).Run(); err != nil {
-			return errors.Errorf("advancing unit life: %w", err)
-		}
-
-		var unitCount entityAssociationCount
-		if err := tx.Query(ctx, lastUnitStmt, unitUUID).Get(&unitCount); errors.Is(err, sqlair.ErrNoRows) {
-			return nil
-		} else if err != nil {
-			return errors.Errorf("getting unit count: %w", err)
-		} else if unitCount.Count != 1 {
-			// The unit is not the last one on the machine.
-			return nil
-		}
-
-		if err := tx.Query(ctx, updateMachineStmt, unitCount).Run(); err != nil {
-			return errors.Errorf("advancing machine life: %w", err)
-		}
+	var unitCount entityAssociationAliveCount
+	if err := tx.Query(ctx, lastUnitStmt, unitUUID).Get(&unitCount); errors.Is(err, sqlair.ErrNoRows) {
+		return "", nil
+	} else if err != nil {
+		return "", errors.Errorf("getting unit count: %w", err)
+	} else if unitCount.AliveCount > 0 {
+		// Nothing to do.
+		return "", nil
+	} else if unitCount.NotAliveCount == 0 {
+		// No units on the machine are marked as dead or dying. If this is the
+		// case then we can assume that the machine is still alive.
+		return "", nil
+	}
 
-		mUUID = unitCount.UUID
+	// We can use the outcome of the update to determine if the machine
+	// was already dying or dead, or if it was successfully advanced to dying.
+	var outcome sqlair.Outcome
+	if err := tx.Query(ctx, updateMachineStmt, entityUUID{UUID: unitCount.UUID}).Get(&outcome); err != nil {
```

Comment:
I think you need to check whether the machine is hosting containers/VMs before doing this.